File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
class Solution {
    public int solution(int[][] triangle) {
        int n = triangle.length;
        int[][] dp = new int[n][n];
        dp[0][0] = triangle[0][0];

        for (int i = 1; i < n; i++) {
            for (int j = 0; j <= i; j++) {
                if (j == 0)
                    dp[i][j] = dp[i-1][j] + triangle[i][j];
                else if (j == i)
                    dp[i][j] = dp[i-1][j-1] + triangle[i][j];
                else
                    dp[i][j] = Math.max(dp[i-1][j-1], dp[i-1][j]) + triangle[i][j];
            }
        }

        int ans = 0;
        for (int j = 0; j < n; j++)
            ans = Math.max(ans, dp[n-1][j]);

        return ans;
    }
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

int arr[501][501];

int solution(vector<vector<int>> triangle) {
    int answer = 0;
    // 그리디가 아니다... 직각삼각형처럼 생각. 
    int n=triangle.size();
    for(int y=n-1; y>=0; y--){
        for(int x=0; x<=y; x++){
            int under=0;
            if(y<n-1) {
                under = max(arr[y+1][x], arr[y+1][x+1]);
            }
            arr[y][x] = under + triangle.at(y).at(x);
        }
    }
    return arr[0][0];
}